### PR TITLE
Warn when a concave polygon is assigned to ConvexPolygonShape2D

### LINKED ISSUE #56382: Check in ConvexPolygonShape2D if convex hull points are used when setting points property
> We could detect in debug builds if a concave shape is assigned to a convex collider and print a warning every time the shape is changed to a concave shape. This check is likely expensive, so it should be disabled in release builds.

See https://github.com/godotengine/godot-docs/issues/5313

### PR DESCRIPTION
**Pull request by Rgtemze**
Fixes #56382

- I've created `left_test` and `is_convex` methods since I couldn't find them in Geometry2D. 
- Every time `_update_shape` is called, the code checks if the polygon (ordered in ccw order) is concave.

### CODE REVIEW

**Review by reduz — APPROVED** [review]

File: scene/resources/convex_polygon_shape_2d.cpp [hunk 1]
```diff
@@ -43,10 +43,38 @@ void ConvexPolygonShape2D::_update_shape() {
...
+		int j = (i + 1) % polygon_size;
+		int k = (j + 1) % polygon_size;
+		// If any consecutive three points fail left-test, then there is a concavity
+		if (!left_test(p_points[i], p_points[j], p_points[k])) {
+			return false;
+		}
+	}
+
+	return true;
+}
```
Comment (akien-mga) [inline scene/resources/convex_polygon_shape_2d.cpp:76]: These should also be in `#ifdef DEBUG_ENABLED`.

Also add dots at the end of comment sentences:
```suggestion
#ifdef DEBUG_ENABLED
// Check if point p3 is to the left of [p1, p2] segment or on it.
bool ConvexPolygonShape2D::left_test(const Vector2 &p1, const Vector2 &p2, const Vector2 &p3) {
	Vector2 p12 = p2 - p1;
	Vector2 p13 = p3 - p1;
	// If the value of the cross product is positive or zero; p3 is to the left or on the segment, respectively.
	return p12.cross(p13) >= 0;
}

bool ConvexPolygonShape2D::is_convex(const Vector<Vector2> &p_points) {
	// Pre-condition: Polygon is in counter-clockwise order.
	int polygon_size = p_points.size();
	for (int i = 0; i < polygon_size && polygon_size > 3; i++) {
		int j = (i + 1) % polygon_size;
		int k = (j + 1) % polygon_size;
		// If any consecutive three points fail left-test, then there is a concavity.
		if (!left_test(p_points[i], p_points[j], p_points[k])) {
			return false;
		}
	}

	return true;
}
#endif
```

File: scene/resources/convex_polygon_shape_2d.h [hunk 2]
```diff
@@ -41,6 +41,8 @@ class ConvexPolygonShape2D : public Shape2D {
 
 protected:
 	static void _bind_methods();
+	static bool left_test(const Vector2 &p1, const Vector2 &p2, const Vector2 &p3);
+	static bool is_convex(const Vector<Vector2> &p_points);
```
Comment (akien-mga) [inline scene/resources/convex_polygon_shape_2d.h:45]: Should be private and not protected I think (or might work also as just static helpers in the .cpp not declared in the header).

**Comment by rburing** [comment]
@Rgtemze Could you implement the requested changes? We'd like to merge this.

**Review by rburing — APPROVED** [review]

**Comment by akien-mga** [comment]
I pushed a rebase doing the suggested changes.
Not critical for 4.0 so we can merge for 4.1.

**Comment by Rgtemze** [comment]
Thank you very much. I'm very sorry I missed the discussion. Is there anything left that I need to do?

**Comment by akien-mga** [comment]
No it's all good, I just need to fix the bug I introduced :P

**Comment by YuriSizov** [comment]
Thanks, and congrats on your first merged pull request!

**Comment by YuriSizov** [comment]
Cherry-picked for 4.0.3.

**Pull request merged**